### ISSUE TITLE
Hash guarantee ({SSHA}AZK)

### ISSUE THREAD

**Issue by Dagon-Project** [Ekultek/Dagon#57]
Dagon has failed to crack given hash: `{SSHA}AZKja92fbuuB9SpRlHqaoXxbTc43Mzc2MDM1Ng==`.

This was attempted on `2017-08-14 22:17:43.384669`..

Commands that were used during failure:
`['dagon.py', '-v', '{SSHA}AZKja92fbuuB9SpRlHqaoXxbTc43Mzc2MDM1Ng==']`

Algorithm attempted: `None`

**Comment by Ekultek** [comment]
Trying to figure out the hash type, give me some time

**Comment by Ekultek** [comment]
Hash type is most likely `ssha-1(base64)`

**Comment by Ekultek** [comment]
hmm.. this should already be implemented:

```
def ssha(string, salt=None, front=False, back=False, **placeholder):
    """
      Create an SSHA hash (seeded salted sha)

      > :param string: string to be hashed
      > :return: a hashed string

      Example:
        >>> ssha("test")
        {SSHA}icUtMxBSzwPv_dSBvwPEwXyK4lo=
    """
    if salt is None:
        salt = os.urandom(4)
    obj = hashlib.sha1(string)
    obj.update(salt)
    if front and not back:
        return "{SSHA}" + base64.urlsafe_b64encode(salt + obj.digest())
    elif back and not front:
        return "{SSHA}" + base64.urlsafe_b64encode(obj.digest() + salt)
    else:
        return "{SSHA}" + base64.urlsafe_b64encode(obj.digest())
```

working on it, give me a few days

**Comment by Ekultek** [comment]
Figured it out, here's the new regex https://regex101.com/r/LtfiTK/1

**Comment by Ekultek** [comment]
Seems like I wasn't encoding the string properly either, so that has been fixed. There is no way I'm going to crack this hash without the salt, I'll give it a shot though

**Comment by Ekultek** [comment]
```
baal@baal-Aspire-5733Z:~/bin/python/dagon$ python dagon.py -v {SSHA}AZKja92fbuuB9SpRlHqaoXxbTc43Mzc2MDM1Ng==
Dagon .. Advanced Hash Manipulation v1.14.32.54(dev)
Clone: https://github.com/ekultek/dagon.git


[*] Starting up at 09:39:50..

[09:39:50 INFO] Analyzing given hash: '{SSHA}AZKja92fbuuB9SpRlHqaoXxbTc43Mzc2MDM1Ng=='...
---------------------------------------------------------------------------
[+] Most Likely Hash Types:
---------------------------------------------------------------------------
[+] SSHA
---------------------------------------------------------------------------

[*] Shutting down at 09:39:50..

baal@baal-Aspire-5733Z:~/bin/python/dagon$ 

```

**Comment by Ekultek** [comment]
```
baal@baal-Aspire-5733Z:~/bin/python/dagon$ python dagon.py -c {SSHA}AZKja92fbuuB9SpRlHqaoXxbTc43Mzc2MDM1Ng== --bruteforce -w /home/baal/bin/python/dagon/downloads/Download-YpcBVaxR.txt
Dagon .. Advanced Hash Manipulation v1.14.32.54(dev)
Clone: https://github.com/ekultek/dagon.git


[*] Starting up at 09:41:03..

[09:41:03 INFO] Reading from, /home/baal/bin/python/dagon/downloads/Download-YpcBVaxR.txt..
[09:41:03 INFO] Found 1 possible hash types to run against: ssha 
[09:41:03 INFO] Starting bruteforce with SSHA..
[09:41:24 WARNING] Unable to find a match for '{SSHA}AZKja92fbuuB9SpRlHqaoXxbTc43Mzc2MDM1Ng==', using SSHA..
[09:41:24 CRITICAL] Ran out of algorithms to try. There are no more algorithms currently available that match this hashes length, and complexity.
[09:41:24 PROMPT] Dagon comes with a hash guarantee, if Dagon is unable to crack your hash successfully, Ekultek will personally attempt to crack your hash for you. Would you like to automatically create a Github issue containing your hash[y/N]: n

[*] Shutting down at 09:41:35..

baal@baal-Aspire-5733Z:~/bin/python/dagon$ python dagon.py -c {SSHA}AZKja92fbuuB9SpRlHqaoXxbTc43Mzc2MDM1Ng== --bruteforce -D downloads
Dagon .. Advanced Hash Manipulation v1.14.32.54(dev)
Clone: https://github.com/ekultek/dagon.git


[*] Starting up at 09:41:47..

[09:41:47 INFO] Reading from, downloads/Download-areNNDEc.txt..
[09:41:47 INFO] Found 1 possible hash types to run against: ssha 
[09:41:47 INFO] Starting bruteforce with SSHA..
[09:42:07 WARNING] Unable to find a match for '{SSHA}AZKja92fbuuB9SpRlHqaoXxbTc43Mzc2MDM1Ng==', using SSHA..
[09:42:07 CRITICAL] Ran out of algorithms to try. There are no more algorithms currently available that match this hashes length, and complexity.
[09:42:07 PROMPT] Dagon comes with a hash guarantee, if Dagon is unable to crack your hash successfully, Ekultek will personally attempt to crack your hash for you. Would you like to automatically create a Github issue containing your hash[y/N]: n
[09:42:44 INFO] Reading from, downloads/Download-VdduplLK.txt..
[09:42:44 INFO] Found 1 possible hash types to run against: ssha 
[09:42:44 INFO] Starting bruteforce with SSHA..
[09:42:45 WARNING] Unable to find a match for '{SSHA}AZKja92fbuuB9SpRlHqaoXxbTc43Mzc2MDM1Ng==', using SSHA..
[09:42:45 CRITICAL] Ran out of algorithms to try. There are no more algorithms currently available that match this hashes length, and complexity.
[09:42:45 PROMPT] Dagon comes with a hash guarantee, if Dagon is unable to crack your hash successfully, Ekultek will personally attempt to crack your hash for you. Would you like to automatically create a Github issue containing your hash[y/N]: n
[09:42:48 INFO] Reading from, downloads/Download-UknLMLOy.txt..
[09:42:48 INFO] Found 1 possible hash types to run against: ssha 
[09:42:48 INFO] Starting bruteforce with SSHA..
[09:42:48 WARNING] Unable to find a match for '{SSHA}AZKja92fbuuB9SpRlHqaoXxbTc43Mzc2MDM1Ng==', using SSHA..
[09:42:48 CRITICAL] Ran out of algorithms to try. There are no more algorithms currently available that match this hashes length, and complexity.
[09:42:48 PROMPT] Dagon comes with a hash guarantee, if Dagon is unable to crack your hash successfully, Ekultek will personally attempt to crack your hash for you. Would you like to automatically create a Github issue containing your hash[y/N]: n
[09:42:49 INFO] Reading from, downloads/Download-jbVUIMWR.txt..
[09:42:49 INFO] Found 1 possible hash types to run against: ssha 
[09:42:49 INFO] Starting bruteforce with SSHA..
[09:42:49 WARNING] Unable to find a match for '{SSHA}AZKja92fbuuB9SpRlHqaoXxbTc43Mzc2MDM1Ng==', using SSHA..
[09:42:49 CRITICAL] Ran out of algorithms to try. There are no more algorithms currently available that match this hashes length, and complexity.
[09:42:49 PROMPT] Dagon comes with a hash guarantee, if Dagon is unable to crack your hash successfully, Ekultek will personally attempt to crack your hash for you. Would you like to automatically create a Github issue containing your hash[y/N]: n
[09:42:50 INFO] Reading from, downloads/Download-sTMtJkAU.txt..
[09:42:50 INFO] Found 1 possible hash types to run against: ssha 
[09:42:50 INFO] Starting bruteforce with SSHA..
[09:42:50 WARNING] Unable to find a match for '{SSHA}AZKja92fbuuB9SpRlHqaoXxbTc43Mzc2MDM1Ng==', using SSHA..
[09:42:50 CRITICAL] Ran out of algorithms to try. There are no more algorithms currently available that match this hashes length, and complexity.
[09:42:50 PROMPT] Dagon comes with a hash guarantee, if Dagon is unable to crack your hash successfully, Ekultek will personally attempt to crack your hash for you. Would you like to automatically create a Github issue containing your hash[y/N]: n
[09:42:51 INFO] Reading from, downloads/Download-qCyKHsNi.txt..
[09:42:51 INFO] Found 1 possible hash types to run against: ssha 
[09:42:51 INFO] Starting bruteforce with SSHA..
[09:42:51 WARNING] Unable to find a match for '{SSHA}AZKja92fbuuB9SpRlHqaoXxbTc43Mzc2MDM1Ng==', using SSHA..
[09:42:51 CRITICAL] Ran out of algorithms to try. There are no more algorithms currently available that match this hashes length, and complexity.
[09:42:51 PROMPT] Dagon comes with a hash guarantee, if Dagon is unable to crack your hash successfully, Ekultek will personally attempt to crack your hash for you. Would you like to automatically create a Github issue containing your hash[y/N]: n
[09:42:51 INFO] Reading from, downloads/Download-NuDjiNln.txt..
[09:42:51 INFO] Found 1 possible hash types to run against: ssha 
[09:42:51 INFO] Starting bruteforce with SSHA..
[09:42:51 WARNING] Unable to find a match for '{SSHA}AZKja92fbuuB9SpRlHqaoXxbTc43Mzc2MDM1Ng==', using SSHA..
[09:42:51 CRITICAL] Ran out of algorithms to try. There are no more algorithms currently available that match this hashes length, and complexity.
[09:42:51 PROMPT] Dagon comes with a hash guarantee, if Dagon is unable to crack your hash successfully, Ekultek will personally attempt to crack your hash for you. Would you like to automatically create a Github issue containing your hash[y/N]: n
[09:42:52 INFO] Reading from, downloads/Download-tIKLRPCt.txt..
[09:42:52 INFO] Found 1 possible hash types to run against: ssha 
[09:42:52 INFO] Starting bruteforce with SSHA..
[09:42:52 WARNING] Unable to find a match for '{SSHA}AZKja92fbuuB9SpRlHqaoXxbTc43Mzc2MDM1Ng==', using SSHA..
[09:42:52 CRITICAL] Ran out of algorithms to try. There are no more algorithms currently available that match this hashes length, and complexity.
[09:42:52 PROMPT] Dagon comes with a hash guarantee, if Dagon is unable to crack your hash successfully, Ekultek will personally attempt to crack your hash for you. Would you like to automatically create a Github issue containing your hash[y/N]: n
[09:42:53 INFO] Reading from, downloads/Download-YpcBVaxR.txt..
[09:42:53 INFO] Found 1 possible hash types to run against: ssha 
[09:42:53 INFO] Starting bruteforce with SSHA..
[09:43:13 WARNING] Unable to find a match for '{SSHA}AZKja92fbuuB9SpRlHqaoXxbTc43Mzc2MDM1Ng==', using SSHA..
[09:43:13 CRITICAL] Ran out of algorithms to try. There are no more algorithms currently available that match this hashes length, and complexity.
[09:43:13 PROMPT] Dagon comes with a hash guarantee, if Dagon is unable to crack your hash successfully, Ekultek will personally attempt to crack your hash for you. Would you like to automatically create a Github issue containing your hash[y/N]: n
[09:43:21 INFO] Reading from, downloads/Download-GtrkaKTT.txt..
[09:43:21 INFO] Found 1 possible hash types to run against: ssha 
[09:43:21 INFO] Starting bruteforce with SSHA..
[09:43:22 WARNING] Unable to find a match for '{SSHA}AZKja92fbuuB9SpRlHqaoXxbTc43Mzc2MDM1Ng==', using SSHA..
[09:43:22 CRITICAL] Ran out of algorithms to try. There are no more algorithms currently available that match this hashes length, and complexity.
[09:43:22 PROMPT] Dagon comes with a hash guarantee, if Dagon is unable to crack your hash successfully, Ekultek will personally attempt to crack your hash for you. Would you like to automatically create a Github issue containing your hash[y/N]: ^C[09:43:23 PROMPT] Dagon comes with a hash guarantee, if Dagon is unable to crack your hash successfully, Ekultek will personally attempt to crack your hash for you. Would you like to automatically create a Github issue containing your hash[y/N]: ^C[09:43:23 CRITICAL] User exited process...
baal@baal-Aspire-5733Z:~/bin/python/dagon$ ^C
```

No way I'm gonna crack this. I'll send a patch for it though

**Comment by Ekultek** [comment]
Fixed via ffd09db
